### DISCUSSION TITLE
Cache clearing in case of errors

### DISCUSSION THREAD

**Discussion by SPARKILATOR** [ecederstrand/exchangelib#950] [Q&A]
I was using this library to connect to the onpremise mailbox, but after 1st retry it was not actually trying to connect to the mailbox server instead it was raising an exception which was stored in a cache.

So to be precise when using the library to connect, errors are cached and not reset in the cache.
As we all know that the CachingProtcol class in the protocol.py is being used here. 

So is there any cache removal logic implemented when exception is cached? because this will raise the exception everytime till the cache is not cleared.

**Comment by ecederstrand** [comment]
That's a good point. There's no logic to automatically evict CachingProtocol cache entries.

When I implemented this I was focused on the retry policy case, where an error creating a Protocol instance is more likely to be a semi-permanent problem and thus not require cache eviction within the duration of the Python process. And in the non-retry case, your script will most likely just exit on errors.

I don't have a very clear picture of what you're trying to attempt, but if you want to retry on connection errors, couldn't you just configure a [retry policy](https://ecederstrand.github.io/exchangelib/#fault-tolerance) and let exchangelib handle the retries for you?

**Reply by SPARKILATOR** [reply]
Thank you for the response.

I just want to say that may be we can have one periodic cache clear logic when it is caching an exception so after that specified time we can able to again connect to the server without restarting any thread.

Because right now it is just trying for the 1st time and then its just raising the exception (not trying to connect to the server) for the next attempts.

So if we add a logic for periodic removal then after the specified time we can again actually try to connect to server.

**Reply by ecederstrand** [reply]
I'd like to understand the use case for this before developing a solution. It's still not clear to me what that use case is:

1. You're writing an ad-hoc script and want immediate feedback on any connection issues to the server. You don't configure a retry policy. An exception is thrown on connection error, your Python process exits and the cache is lost.
2. You want to keep trying until either a connection is established or you give up permanently. You configure a retry policy to retry for as long as you specify. Eventually an exception is thrown on connection error. You catch the exception. The failure is cached, which is fine because the connection error is supposedly permanent.

Those two scenarios are already covered, I believe. What other scenarios do we have that require different behavior?

I can live with not caching the error in the ad-hoc case. In the latter case, it's not clear to me what logic to apply before removing the cache entry. Let's say you tried an endpoint for 1 hour without success. When do you retry? After 1 hour? A day? A week?
